Sources/SwiftlyCore/HTTPClient.swift

```diff
@@ -179,30 +137,60 @@ public struct SwiftlyHTTPClient {
         public let url: String
     }
 
-    public func downloadToolchain(
-        _ toolchain: ToolchainVersion,
-        url: String,
-        to destination: String,
-        reportProgress: @escaping (DownloadProgress) -> Void
-    ) async throws {
-        try await self.downloader.downloadToolchain(
-            toolchain,
-            url: url,
-            to: destination,
-            reportProgress: reportProgress
-        )
+    public func downloadFile(url: URL, to destination: URL, reportProgress: @escaping (DownloadProgress) -> Void) async throws {
+        let fileHandle = try FileHandle(forWritingTo: destination)
+        defer {
+            try? fileHandle.close()
+        }
+
+        let request = makeRequest(url: url.absoluteString)
+        let response = try await self.inner.execute(request, timeout: .seconds(30))
+
+        guard case response.status = HTTPResponseStatus.ok else {
+            throw Error(message: "Received \(response.status) when trying to download \(url)")
+        }
+
+        // Unknown download.swift.org paths redirect to a 404 page which then returns a 200 status.
+        // As a heuristic for if we've hit the 404 page, we check to see if the content is HTML.
+        if url.host == "download.swift.org" {
+            guard !response.headers["Content-Type"].contains(where: { $0.contains("text/html") }) else {
+                throw SwiftlyHTTPClient.DownloadNotFoundError(url: url.path)
```

Comment:
I was going to ask can't you check the response status, but apparently not. The 404 page returns a status of 302. You could check the location header is `https://swift.org/404.html` instead though

Reply:
So after following the redirect, swift.org/404.html does in fact return a status code of 404 nowadays, so I've just gone ahead and removed this special check altogether.